### ISSUE TITLE
Difference between JsonUtils.classpathToObject and ClassLoader.getResourceAsStream?

### ISSUE THREAD

**Issue by efenderbosch** [bazaarvoice/jolt#265]
I've got a simple test controller:

``` java
    public Result transform(String applicationName, String entityType) throws Exception {
        String specPath = "jolt/" + applicationName + "/" + entityType + "-spec.json";
        InputStream is = this.getClass().getClassLoader().getResourceAsStream(specPath);
        // Object spec = JsonUtils.classpathToObject(specPath);
        Object spec = JsonUtils.jsonToObject(is);
        Chainr chainr = Chainr.fromSpec(spec);
        String input = request().body().asJson().toString();
        Object transformed = chainr.transform(JsonUtils.jsonToObject(input));
        String json = JsonUtils.toJsonString(transformed);
        return ok(json);
    }
```

If I use get the `InputStream` from the `ClassLoader` read it with `JsonUtils.jsonToObject`  it works correctly.

If I try to use `JsonUtils.classpathToObject` then the following exception is thrown:

```
Caused by: java.lang.RuntimeException: Unable to load JSON object from InputStream.
        at com.bazaarvoice.jolt.JsonUtilImpl.jsonToObject(JsonUtilImpl.java:105)
        at com.bazaarvoice.jolt.JsonUtilImpl.classpathToObject(JsonUtilImpl.java:198)
        at com.bazaarvoice.jolt.JsonUtils.classpathToObject(JsonUtils.java:157)
        at com.company.TestTransformingController.transform(TestTransformingController.java:19)
        at router.Routes$$anonfun$routes$1$$anonfun$applyOrElse$1$$anonfun$apply$1.apply(Routes.scala:73)
        at router.Routes$$anonfun$routes$1$$anonfun$applyOrElse$1$$anonfun$apply$1.apply(Routes.scala:73)
        at play.core.routing.HandlerInvokerFactory$$anon$4.resultCall(HandlerInvoker.scala:157)
        at play.core.routing.HandlerInvokerFactory$$anon$4.resultCall(HandlerInvoker.scala:156)
        at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$14$$anon$3$$anon$1.invocation(HandlerInvoker.scala:136)
        at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:73)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: No content to map due to end-of-input
 at [Source: UNKNOWN; line: 1, column: 0]
        at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:255)
        at com.fasterxml.jackson.databind.ObjectMapper._initForReading(ObjectMapper.java:3851)
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3792)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2867)
        at com.bazaarvoice.jolt.JsonUtilImpl.jsonToObject(JsonUtilImpl.java:102)
        at com.bazaarvoice.jolt.JsonUtilImpl.classpathToObject(JsonUtilImpl.java:198)
        at com.bazaarvoice.jolt.JsonUtils.classpathToObject(JsonUtils.java:157)
```

Am I doing something wrong, or is this behavior expected?

Thanks

**Comment by efenderbosch** [comment]
I'm having another ClassLoader issue.

I can use `Class.forName("com.company.MyCustomTransform")` and the class is found just fine.

If I put it in a spec:

```json
[
  {
    "operation": "com.company.MyCustomTransform"
  }
]
```

Then I get a CNFE:

`
Caused by: java.lang.ClassNotFoundException: com.company.MyCustomTransform
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at com.bazaarvoice.jolt.chainr.spec.ChainrEntry.loadJoltTransformClass(ChainrEntry.java:134)
        at com.bazaarvoice.jolt.chainr.spec.ChainrEntry.<init>(ChainrEntry.java:104)
        at com.bazaarvoice.jolt.chainr.spec.ChainrSpec.<init>(ChainrSpec.java:68)
        at com.bazaarvoice.jolt.chainr.ChainrBuilder.<init>(ChainrBuilder.java:40)
        at com.bazaarvoice.jolt.Chainr.fromSpec(Chainr.java:92)
`

We are using the Play Framework.

**Comment by milosimpson** [comment]
My suspicion is that there are issues with ClassLoaders and how you are deploying / using the system.

When the Jolt code is doing the Class.forName(), my suspicion is that it is using a ClassLoader that is "somehow" limited to the Jolt Jar file, and therefore can't find your .class files.

**Comment by efenderbosch** [comment]
I think you're right. When I run the code in a JUnit test it works just fine. Inside a Play Framework Controller it doesn't. The Play Framework is using `play.runsupport.DelegatedResourcesClassLoader`. Not sure how I can fix this.

**Comment by efenderbosch** [comment]
Yep, that's it. Inside my controller:

```java
\\ the following line works
this.getClass().getClassLoader().loadClass("com.company.MyCustomTransform");
\\ the following line throws CNFE
Chainr.class.getClassLoader().loadClass("com.company.MyCustomTransform");
```

**Comment by milosimpson** [comment]
So you can implement your own "ChainrInstantiator" and use the fromSpec method that takes one.

```
    public static Chainr fromSpec( Object input, ChainrInstantiator instantiator ) {
        return new ChainrBuilder( input ).loader( instantiator ).build();
    }
```

You would basically just copy the DefaultChainrInstantiator but use your class to do the loading.

Note, the purpose of the "ChainrInstantiator" and friends code, was not to work around classLoading issues, but to allow for users to use Guice to load classes (which we do).

**Comment by efenderbosch** [comment]
Will that work? ChainrEntry.operationClassName is private, so I can't get to it to load using a different ClassLoader.

**Comment by milosimpson** [comment]
Herm, yeah.    ChainrEntry does the classForName, and the ChainrInstantiator calls the constructor.

Dunno.

**Comment by efenderbosch** [comment]
The only solution that would really be clean would be to add another `Chainr.fromSpec(Object input, ClassLoader classLoader)` method that passes that `ClassLoader` around, defaulting to `ClassLoader.getSystemClassLoader()` if not provided.

Maybe I could fork, add that, and submit a pull request?

**Comment by efenderbosch** [comment]
Any easier solution might be to use `Thread.currentThread().getContextClassLoader().loadClass(String className)` instead of `Class.forName(String className)`.

**Comment by milosimpson** [comment]
I am more on board with the "Thread.currentThread().getContextClassLoader().loadClass(String className)" fix, if it works : minimal change and don't have to pipe a ClassLoader around the place.

**Comment by efenderbosch** [comment]
Sorry to keep spamming, but after a bit of reading it seems like the solution proposed here on StackOverflow would work and be fairly non-intrusive:

http://stackoverflow.com/a/36228195

TL;DR: use `getClass().getClassLoader().loadClass()`

**Comment by milosimpson** [comment]
That works for me as well.   

You should fork and do the build, to make sure it works with your Play framework stuff.

**Comment by efenderbosch** [comment]
Hmmm... that didn't work, but `Thread.currentThread().getContextClassLoader()` does. I'll continue on this tomorrow.

**Comment by efenderbosch** [comment]
Another option might be to expose a library-scoped `ClassLoader`. Something like


```java
public class Chainr implements Transform, ContextualTransform {
    // ...
    public static ClassLoader Chainr.CLASS_LOADER = Chainr.class.getClassLoader();
    // ...
}
```

This would get rid of the need to add a `ClassLoader` to constructors and carrying that value around between all the various classes.

And used like this:

```java
public class ChainrEntry {
    // ...
    private Class<? extends JoltTransform> loadJoltTransformClass() {
        // ...
        Class opClass = Chainr.CLASS_LOADER.loadClass( operationClassName );
        // ...
    }
    // ...
}
```

It could then be overridden as necessary:

```java

public class MyController extends Controller {

    static {
        Chainr.CLASS_LOADER = MyController.class.getClassLoader();
    }
    // ...
}
```

The problem here is that the json-utils module would need to add the jolt-core project to its dependencies so that `JsonUntilImpl` could replace `InputStream inputStream = this.getClass().getResourceAsStream( classPath );` with `InputStream inputStream = Chainr.CLASS_LOADER.getResourceAsStream( classPath );`.

**Comment by efenderbosch** [comment]
Nope... that makes the unit tests in json-utils fail.

**Comment by milosimpson** [comment]
Yeah, having had a night to sleep on it.   I think the best thing is to add a .withClassLoader() to the ChainrBuilder object, and then pass it down as needed.

Its a reasonable thing to add to the Builder; the piping down to the ChainrEntry is ugly, but can be cleaned up / refactored later.
